egressgateway: Add IPv6 support for egress IP

syedazeez337 opened the pull request:
## Description
This PR adds support for setting IPv6 addresses as egress IPs in CiliumEgressGatewayPolicy resources. Previously, only IPv4 addresses were supported.

The changes include:
- Updating the CRD definition to accept both IPv4 and IPv6 addresses
- Modifying the policy parsing code to handle IPv6 egress IPs
- Updating the documentation to reflect the new IPv6 support
- Adding tests to verify the implementation

## Related issues
Fixes: #38962

## How has this been tested
Added unit tests to verify IPv6 egress gateway policy parsing and configuration.
Built the Cilium agent to ensure compatibility with the existing codebase.

## Types of changes
- Feature: Added IPv6 support for egress gateway policies

## Checklist
- [x] I have read the [Documentation](https://docs.cilium.io)
- [x] I have read the [Contributing documentation](https://docs.cilium.io/en/latest/contributing/development/)
- [x] I have signed the [CLA](https://cla.docs.cilium.io/)
- [x] I have run `make check-copyright` to check that the copyright headers are valid
- [x] I have run `make lint` to check the code style
- [x] I have run `make -C Documentation spelling` to check for misspellings
- [x] I have added tests to cover my changes
- [x] I have updated the documentation accordingly

rgo3 commented on .github/release-notes/egress-gateway-ipv6.yaml:
```diff
@@ -0,0 +1,4 @@
+kind: feature
```
I don't think we usually use yaml files for release notes AFAIK, we can probably remove this file.

rgo3 commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -93,9 +93,11 @@ func (r *gatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
-	if err := r.Client.List(ctx, tlsRouteList); err != nil {
-		scopedLog.ErrorContext(ctx, "Unable to list TLSRoutes", logfields.Error, err)
-		return r.handleReconcileErrorWithStatus(ctx, err, original, gw)
+	if helpers.HasTLSRouteSupport(r.Client.Scheme()) {
```
The gateway-api related changes seem unrelated to the egress gateway changes you are making, better to split that into a separate PR.

rgo3 commented on pkg/egressgateway/manager.go:
```diff
@@ -185,8 +185,8 @@ func NewEgressGatewayManager(p Params) (out struct {
 
-	// TODO: deprecate --enable-ipv4-egress-gateway, create new --enable-egress-gateway
-	if !dcfg.EnableIPv4EgressGateway {
+	// Check if either the new flag or the deprecated flag is enabled
```
I think deprecating the old and introducing the new flag could also be split into a separate PR.

syedazeez337 replied:
Let me look into your comments and rectify my code to suite the issue. Thank you and I will keep updating.

syedazeez337 committed:
egressgateway: Add IPv6 support

This commit adds IPv6 support to the egress gateway feature:

- Add a new --enable-egress-gateway flag that enables egress gateway for both IPv4 and IPv6
- Mark the existing --enable-ipv4-egress-gateway flag as deprecated
- Update the validation format for EgressIP from ipv4 to ip to support both IPv4 and IPv6 addresses
- Modify the code to handle IPv6 egress IPs
- Add IPv6 constants for excluded CIDRs, gateway not found, and egress IP not found
- Fix a bug in the updateEgressRules6 function where it was using IPv4 constants for IPv6 excluded CIDRs
- Update the Helm chart to use the new flag
- Update documentation to reflect the new IPv6 support and add examples for IPv6 egress gateway policies
- Update metrics and cilium-cli feature detection to check for both flags
- Update BPF code to use both flags

Signed-off-by: Syed Azeez <syedazeez337@gmail.com>

rgo3 reviewed (changes requested):
Hey @syedazeez337, thank you very much for this PR! Especially already tackling the documentation changes, this was still on my TODO list.

It looks like this PR currently includes a mix of slightly unrelated changes, could you please split those into separate PRs? I see potentially 3 different PRs here: ipv6 support for egress IP, introducing `--enable-egress-gateway` and the gateway-api fix you included here. Some of my inline comments refer to making this cleanup as well.

I've only skimmed the egressIP changes, and so far they look good. However, I think it would be nice that we do certain checks when parsing egress gateway policies that have the egress IP field set. E.g., in practice we should reject policies that set a ipv4 egress IP, but include ipv6 destination CIDRs and vice versa.

syedazeez337 committed:
Merge branch 'cilium:main' into main

syedazeez337 committed:
egressgateway: Add IPv6 support for egress IP

This commit adds support for setting IPv6 addresses as egress IPs in
CiliumEgressGatewayPolicy resources. Previously, only IPv4 addresses were supported.

The changes include:
- Updating the CRD definition to accept both IPv4 and IPv6 addresses
- Modifying the policy parsing code to handle IPv6 egress IPs
- Adding validation to reject policies with mismatched IP families
- Updating the documentation to reflect the new IPv6 support
- Adding tests to verify the implementation

Fixes: #38962
Signed-off-by: Syed Azeez <syedazeez337@gmail.com>
